tests/cfg.c

```diff
@@ -0,0 +1,266 @@
+#undef NDEBUG
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+
+#include <security/pam_appl.h>
+
+#include "cfg.h"
+
+static void write_conf(int fd, const char *text) {
+  size_t len;
+  int e;
+
+  len = strlen(text);
+  while (len) {
+    ssize_t w;
+
+    w = write(fd, text, len);
+    assert(w >= 0);
+
+    len -= w;
+    text += w;
+  }
+
+  e = fsync(fd);
+  assert(e == 0);
+}
+
+static char *generate_conf_file_arg(void) {
+  // Generate a conf_file= argument
+  //
+  // The function returns a string which is:
+  // - suitable as argv item for pam_u2f
+  // - suitable as template argument for mkstemp
+  // - referring to the absolute path of the temporary file.
+
+  char cwd[1024];
+  char *template;
+  int err;
+
+  err = !getcwd(cwd, sizeof(cwd));
+  assert(!err);
+
+  err = asprintf(&template, "conf_file=%.*s/test_config_XXXXXX",
+                 (int) sizeof(cwd), cwd) == -1;
+  assert(!err);
+
+  return template;
+}
+
+static void test_regular(void) {
+  // Testing regular behaviour:
+  // - config file inheritance: the config file provides defaults,
+  //   overrides from regular argv
+  // - comments
+  // - indentation
+
+  char *config_arg;
+  int fd, e;
+  cfg_t cfg;
+
+  config_arg = generate_conf_file_arg();
+
+  fd = mkstemp(config_arg + strlen("conf_file="));
+  assert(fd != -1);
+
+  write_conf(fd,
+             "max_devices=10\n"
+             "manual\n"
+             "debug\n"
+             "nouserok\n"
+             "openasuser\n"
+             "alwaysok\n"
+             "interactive\n"
+             "cue\n"
+             "nodetect\n"
+             "expand\n"
+             "userpresence=0\n"
+             "userverification=0\n"
+             "pinverification=0\n"
+             "authfile=/foo/bar\n"
+             "# sshformat\n"                 // Commented -> no effect
+             "  # sshformat\n"               // also commented
+             " authpending_file=/baz/quux\n" // Indented -> accepted
+             "origin=pam://lolcalhost\n"
+             "appid=pam://lolcalhost\n"
+             "prompt=hello\n"
+             "cue_prompt=howdy\n"
+             "debug_file=stdout\n");
+
+  const char *argv[] = {
+    config_arg, "prompt=hi",
+    "debug", // So we have a log file for the test
+  };
+
+  e = cfg_init(&cfg, 0, sizeof(argv) / sizeof(*argv), argv);
+  assert(e == PAM_SUCCESS);
+
+  unlink(config_arg + strlen("conf_file="));
+  free(config_arg);
+  close(fd);
+
+  assert(cfg.alwaysok);
```

Comment:
There's a lot of settings set in the config "file" above that are not tested for below.  Should we assert those also or is that overkill?

Reply:
My intention here was to test at least one integer value, and any string values (with the latter being more _risky_).

I like the idea of extending the test to cover all values. I suppose the test would require to flip over all default values, and observe how they change.